flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/async/queue/UnorderedStreamElementQueue.java

```diff
@@ -19,287 +19,261 @@
+import org.apache.flink.streaming.api.watermark.Watermark;
+import org.apache.flink.streaming.runtime.streamrecord.StreamElement;
+import org.apache.flink.streaming.runtime.streamrecord.StreamRecord;
 import org.apache.flink.util.Preconditions;
 
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 import java.util.ArrayDeque;
-import java.util.Arrays;
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Deque;
 import java.util.HashSet;
-import java.util.Iterator;
+import java.util.List;
+import java.util.Optional;
+import java.util.Queue;
 import java.util.Set;
-import java.util.concurrent.Executor;
-import java.util.concurrent.locks.Condition;
-import java.util.concurrent.locks.ReentrantLock;
 
 /**
- * Unordered implementation of the {@link StreamElementQueue}. The unordered stream element queue
- * emits asynchronous results as soon as they are completed. Additionally it maintains the
- * watermark-stream record order. This means that no stream record can be overtaken by a watermark
- * and no watermark can overtake a stream record. However, stream records falling in the same
- * segment between two watermarks can overtake each other (their emission order is not guaranteed).
+ * Unordered implementation of the {@link StreamElementQueue}. The unordered stream element queue provides
+ * asynchronous results as soon as they are completed. Additionally, it maintains the watermark-stream record order.
+ *
+ * <p>Elements can be logically grouped into different stages separated by watermarks. A stage needs to be
+ * completely emitted before entries from a following stage are emitted. Thus, no stream record can be overtaken
+ * by a watermark and no watermark can overtake a stream record.
+ * However, stream records falling in the same stage between two watermarks can overtake each other (their emission
+ * order is not guaranteed).
  */
 @Internal
-public class UnorderedStreamElementQueue implements StreamElementQueue {
+public final class UnorderedStreamElementQueue<OUT> implements StreamElementQueue<OUT> {
 
 	private static final Logger LOG = LoggerFactory.getLogger(UnorderedStreamElementQueue.class);
 
 	/** Capacity of this queue. */
 	private final int capacity;
 
-	/** Executor to run the onComplete callbacks. */
-	private final Executor executor;
-
-	/** OperatorActions to signal the owning operator a failure. */
-	private final OperatorActions operatorActions;
-
-	/** Queue of uncompleted stream element queue entries segmented by watermarks. */
-	private final ArrayDeque<Set<StreamElementQueueEntry<?>>> uncompletedQueue;
-
-	/** Queue of completed stream element queue entries. */
-	private final ArrayDeque<StreamElementQueueEntry<?>> completedQueue;
-
-	/** First (chronologically oldest) uncompleted set of stream element queue entries. */
-	private Set<StreamElementQueueEntry<?>> firstSet;
+	/** Queue of queue entries segmented by watermarks. */
+	private final Deque<Stage<OUT>> stages;
```

Comment:
Personally I find "segment" as a more appropriate word here (also was used in the old java doc here) compared to word "stage".

Reply:
Okay, will change.